LSPS5/README.md

```diff
@@ -0,0 +1,264 @@
+
+| Name    | `webhook_registration` |
+| ------- | ---------------------- |
+| Version | 1                      |
+| Status  | For Implementation     |
+
+## Motivation
+
+In typical mobile environments, a program that is not currently
+being focused on by the user will be suspended, with all its
+TCP connections dropped.
+
+Typically, such a program would not be able to get either CPU
+time or any ability to receive information from network.
+
+If an LSPS client is implemented as a program on such a mobile
+environment, then when it is suspended, the LSP cannot contact
+the client to perform any operations, such as to forward a
+payment to the client.
+
+Generally, mobile environments will have the mobile OS developers
+run a server or network of servers where an application developer
+can deliver "push notifications".
+When an application developer server contacts the mobile OS
+developer push notification server, the application developer
+server signs and sends a message that is to be displayed to the
+user of the mobile unit, which the mobile OS server will validate
+before causing the message to be displayed to the specified user.
+If the user then taps on this "push notification" on their screen,
+the application is awoken and it can then re-establish its TCP
+connections.
+
+There may be multiple implementations of mobile LSPS-supporting
+clients, each of which run their own application developer server
+and have their own signing keys for these push notifications.
+
+This specification provides a way for a mobile client to
+register some specific webhook by which the LSP can signal a
+push notification to the application developer server, which
+will in turn convert the push notification to one that it itself
+signs and can send to the mobile OS developer server.
+
+### Actors
+
+The 'client' is a specific instance of the client application
+running on some mobile device.
+
+The 'LSP' is a published node that may need to awaken some
+client instance in order to handle some event.
+
+The 'client application server' is the server that a mobile
+client application developer maintains, which is contacted by
+the LSP via an HTTPS `POST` request, in order for the LSP
+to awaken some client instance.
+
+The 'mobile OS server' is the entity that actually allows
+push notifications to be sent to actual mobile devices.
+
+## Protocol
+
+### Version Negotiation
+
+The `lsps5.getversion` call is used by the client to determine
+which versions of this API the LSP supports.
+It takes no parameters `{}` and has no defined errors.
+
+The `lsps5.getversion` call results in:
+
+```JSON
+{
+   "versions": [1]
+}
+```
+
+`versions` is an array of versions of this API that the LSP
+supports.
+
+### Webhook Registration API
+
+The client can register a specific webhook for the LSP to
+contact in case the client is asleep, but the LSP needs the
+client to be online and ready to interact.
+
+The client uses the `lsps5.setwebhook` call to specify the
```

Comment:
not sure if this is worth mentioning, but it's probably OK for the client app server to make this call too?

Reply:
The LSPS spec is intended to build on top of LSPS0, where all communications is over LN protocol via BOLT8 message ID 37913.  See LSPS0 for why it is done that way.

This has the advantage that the LSP can identify the client node ID; establishing a BOLT8 tunnel requires knowledge of the client node ID private key.  If you are OK with the client app server having that private key, that is fine, but then your client app server is able to do things like send `update_add_htlc` in behalf of the client, i.e. custodial.

Reply:
ah, I see that makes sense. I was aware that LSPS0 uses LN messages, but forgot about BOLT 8 tunnel establishments requiring the private key.

we were initially hoping to allow mobile clients to delegate calls **to** LSP using a proxy API service, while clients maintain their own separate BOLT 8 tunnel with LSP as a data plane (so when the client wakes, they could still do `update_add_htlc` themselves, for example).

This model has a couple of benefits:
1. We don't need to rely on the availability of mobile bindings for [ldk-lsp-client](https://github.com/lightningdevkit/ldk-lsp-client) to use this (and LSPS2) right away. And being able to run things in Rust outright is nice.
2. If LSP implements some kind of authentication with API keys, we want to be able to avoid shipping each version binary with the API key, since rotating API keys will effectively require us to force everyone to upgrade their app.

So, from the sound of it, since we're using LDK, we'll need to wait for both the client and the bindings to become available before mobile devices can use all the things in LSPS spec?

Reply:
Yes, looks like it.

Currently LSPS0 has no provision for API keys.  Do we need a spec for that?

I was thinking you could have the initial NOISE handshake be done in coordination with the client and the client can hand over the session keys to the proxy API, but then the LSP would have two connections from the same apparent node ID, and most LN software (including LDK) maintain a "one connection per peer node ID" rule.  And it still allows the proxy API server to spoof `update_add_htlc`.

Reply:
How are API keys issued?  If the client is open-sourced, what prevents me from just copying the API keys from the open-source client to my own bespoke client (or conversely, from forking your client and adding more stuff that you might not want your client to be blamed for?).  If the API keys are issued with some strong check that the client is a 100% bonafide unmodified binary from you, then it kinda loses the point of open-sourcing the client in the first place (yeah it is open source but you still cannot modify it in practice since the LSP will reject you anyway).